### PR TITLE
Fixes #7207: Select and display agent mode (verify/enforce)

### PR DESCRIPTION
**Pull request by RaphaelGauthier**
https://www.rudder-project.org/redmine/issues/7207

### CODE REVIEW

File: rudder-web/src/main/scala/com/normation/rudder/web/components/DirectiveEditForm.scala [hunk 1]
```diff
@@ -427,6 +430,59 @@ class DirectiveEditForm(
...
+        If set to <b>Enforce</b> this directive will appply necessary changes except on Nodes with a <b>Verify</b> override setting.
+      </div>
+    case false =>
+      <p>
+        Currrent global settings do not allow this mode to be overriden on a per-directive bases. You may change this in <b>Settings</b>,
+        or contact your Rudder administrator about this.
+      </p>
+  }
+    private[this] val policyModes = {
+      val globalPolicyModeName = configService.rudder_policy_mode_name().getOrElse("") match {
```
Comment (VinceMacBuche) [inline rudder-web/src/main/scala/com/normation/rudder/web/components/DirectiveEditForm.scala]: globalPolicyMode should be passed as parameter of DirectiveEditForm

File: rudder-web/src/main/scala/com/normation/rudder/web/components/DirectiveEditForm.scala [hunk 2]
```diff
@@ -427,6 +430,59 @@ class DirectiveEditForm(
...
+        case Verify  => "Audit"
```
Comment (VinceMacBuche) [inline rudder-web/src/main/scala/com/normation/rudder/web/components/DirectiveEditForm.scala]: You should not rename modes to Audit or Enforce, you should do "policyMode.name.capitalize()"

File: rudder-web/src/main/scala/com/normation/rudder/web/components/DirectiveEditForm.scala [hunk 3]
```diff
@@ -427,6 +430,59 @@ class DirectiveEditForm(
...
+        case Enforce => "Enforce"
+      }
+      val policyName = directive.policyMode match {
+        case Some(Verify) => "Audit"
```
Comment (VinceMacBuche) [inline rudder-web/src/main/scala/com/normation/rudder/web/components/DirectiveEditForm.scala]: Same here Should use mode.name

File: rudder-web/src/main/scala/com/normation/rudder/web/components/DirectiveEditForm.scala (hunk 1 above)
Comment (VinceMacBuche) [inline rudder-web/src/main/scala/com/normation/rudder/web/components/DirectiveEditForm.scala]: At least somewhere else where you managed a possible error

File: rudder-web/src/main/scala/com/normation/rudder/web/components/DirectiveEditForm.scala [hunk 4]
```diff
@@ -427,6 +430,59 @@ class DirectiveEditForm(
     }
     s"${version} ${deprecationInfo}"
   }
+  private[this] val globalOverrideText = configService.rudder_policy_overridable().getOrElse("") match {
```
Comment (VinceMacBuche) [inline rudder-web/src/main/scala/com/normation/rudder/web/components/DirectiveEditForm.scala]: Same here: globalPolicyMode should be passed as parameter of DirectiveEditForm

File: rudder-web/src/main/scala/com/normation/rudder/web/services/ReportDisplayer.scala [hunk 5]
```diff
@@ -98,13 +101,7 @@ class ReportDisplayer(
   def asyncDisplay(node : NodeInfo) : NodeSeq = {
     val id = JsNodeId(node.id)
     val callback =  SHtml.ajaxInvoke(() => SetHtml("reportsDetails",displayReports(node)) )
-    Script(OnLoad(JsRaw(
```
Comment (VinceMacBuche) [inline rudder-web/src/main/scala/com/normation/rudder/web/services/ReportDisplayer.scala]: Why did you remove the tab "behavior" ?

File: rudder-web/src/main/webapp/javascript/rudder/angular/auditMode.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
+
+var app = angular.module('auditmode', []);
+
+app.factory('configGlobalFactory', function ($http){
+  //Case : global configuration
+  this.policymode = {
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/auditMode.js]: policyMode

File: rudder-web/src/main/webapp/javascript/rudder/angular/auditMode.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+ , save      : function(mode){
+	             var data = {'value' : mode};
+	             return $http.post(this.url, data).then(function successCallback(response) {
+	               return response.status==200;
+	             }, function errorCallback(response) {
+	               return response.status==200;
+	             });
+	           }
+  };
+  this.overridemode = {
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/auditMode.js]: overrideMode

File: rudder-web/src/main/webapp/javascript/rudder/angular/auditMode.js [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+                   return response.status==200;
+                 });
+               }
+  };
+  return this;
+});
+
+app.factory('configNodeFactory', function ($http){
+  //Case : node configuration
+  this.policymode = {
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/angular/auditMode.js]: policyMode

File: rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js [hunk 9]
```diff
@@ -273,7 +271,7 @@ function createRuleTable(gridId, data, needCheckbox, needActions, needCompliance
           var tooltipId = data.id+"-description";
           parent.attr("tooltipid",tooltipId);
           parent.attr("title","");
-          parent.addClass("tooltip tooltipabletr");
+          parent.addClass("tooltip tooltipabletr");{}
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js]: you can remove the {}

File: rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js [hunk 10]
```diff
@@ -706,8 +705,9 @@ function createDirectiveTable(isTopLevel, isNodeView, contextPath) {
           editLink.click(function(e) {e.stopPropagation();});
           editLink.append(editIcon);
           editLink.addClass("reportIcon");
-
           $(nTd).append(editLink);
+          var policyMode = oData.policyMode ? oData.policyMode : "";
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js]: Maybe you should keep undefined / treat undefined (find it better to deal with than empty String

File: rudder-web/src/main/webapp/javascript/rudder/rudder.js [hunk 11]
```diff
@@ -606,4 +589,74 @@ $(document).ready(function() {
...
 });
 
+function createTextAgentPolicyMode(isNode, currentPolicyMode, explanation){
+  var policyMode = currentPolicyMode.toLowerCase();
+  var nodeOrDirective = isNode ? "node" : "directive";
+  var labelType = "label-"+policyMode;
+  var policyName = policyMode == "verify" ? "audit" : policyMode
+  var span = "<span class='label-text " + labelType + " glyphicon glyphicon-question-sign' data-toggle='tooltip' data-placement='top' data-html='true' title=''></span>"
+  var badge = $(span).get(0);
+  var tooltip = "" +
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/rudder.js]: I think the tooltip can be put in an external function ( like : 'policyModetooltip (kind, policyName, explanation)'

File: rudder-web/src/main/webapp/javascript/rudder/rudder.js [hunk 12]
```diff
@@ -606,4 +589,74 @@ $(document).ready(function() {
   createTooltip();
 });
 
+function createTextAgentPolicyMode(isNode, currentPolicyMode, explanation){
+  var policyMode = currentPolicyMode.toLowerCase();
+  var nodeOrDirective = isNode ? "node" : "directive";
+  var labelType = "label-"+policyMode;
+  var policyName = policyMode == "verify" ? "audit" : policyMode
```
Comment (VinceMacBuche) [inline rudder-web/src/main/webapp/javascript/rudder/rudder.js]: You should use policyMode everywhere, not policyName

File: rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js (hunk 10 above)
Comment (RaphaelGauthier) [inline rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js]: Right !

File: rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js [hunk 13]
```diff
@@ -769,17 +770,17 @@ function createNodeComplianceTable(gridId, data, contextPath, refresh) {
...
         var editLink = $("<a />");
         editLink.attr("href",contextPath +'/secure/nodeManager/searchNodes#{"nodeId":"'+oData.id+'"}');
         var editIcon = $("<img />");
         editIcon.attr("src",contextPath + "/images/icMagnify-right.png");
         editLink.click(function(e) {e.stopPropagation();});
         editLink.append(editIcon);
         editLink.addClass("reportIcon");
-
         $(nTd).append(editLink);
+        console.log(oData)
```
Comment (RaphaelGauthier) [inline rudder-web/src/main/webapp/javascript/rudder/rudder-datatable.js]: Arf :)

File: rudder-web/src/main/scala/com/normation/rudder/web/services/ReportDisplayer.scala (hunk 5 above)
Comment (RaphaelGauthier) [inline rudder-web/src/main/scala/com/normation/rudder/web/services/ReportDisplayer.scala]: I remember now. This code doesn't work since the bind function is deprecated (since Jquery 3.0.0) !

**Comment by RaphaelGauthier** [comment]
PR updated

**Comment by VinceMacBuche** [comment]
OK, merging this PR
